scipy/stats/stats.py

```diff
@@ -408,7 +411,20 @@ def gmean(a, axis=0, dtype=None):
             log_a = np.log(np.asarray(a, dtype=dtype))
     else:
         log_a = np.log(a)
-    return np.exp(log_a.mean(axis=axis))
+
+    if weights:
+        weights = np.array(weights, dtype=dtype, copy=False)
+        elif dtype:
+            # Must change the default dtype allowing array type
+            if isinstance(a, np.ma.MaskedArray):
+                weights = np.ma.asarray(weights, dtype=dtype)
+            else:
+                weights = np.asarray(weights, dtype=dtype)
+
+        if weights.shape != log_a.shape:
```

Comment:
Thanks for the `ValueError`.
On more carefuly consideration, I don't think the shapes need to be identical. I think `np.average` has the size checking you need. I would actually suggest not raising your own message because the message `np.average` provides is quite readable, e.g.
```
In [25]: np.exp(np.average(log_a, axis=0, weights=weights))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-25-612396c6a193> in <module>
----> 1 np.exp(np.average(log_a, axis=0, weights=weights))

<__array_function__ internals> in average(*args, **kwargs)

~\Anaconda3\envs\scipydev\lib\site-packages\numpy\lib\function_base.py in average(a, axis, weights, returned)
    399             if wgt.shape[0] != a.shape[axis]:
    400                 raise ValueError(
--> 401                     "Length of weights not compatible with specified axis.")
    402
    403             # setup wgt to broadcast along axis

ValueError: Length of weights not compatible with specified axis.
```

Reply:
Makes sense. Resolved in d01121656975514443a25aae9221b29db403ebcd.